### ISSUE TITLE
Some projects don't complete, then savegame loss

### ISSUE THREAD

**Issue by a-horde-of-cats** [frnsys/the_founder#159]
Later in the game, when I had 7 simultaneous projects, sometimes projects would stop having a pop up window on conclusion. This sometimes resolved after restarting or deleting all projects.

Also, on one restart, the save file was lost.

**Comment by frnsys** [comment]
Sorry to hear that--did you happen to see any error messages or anything like that? Or when you try to reload your game?

**Comment by a-horde-of-cats** [comment]
No error messages. Products would complete without any window popping up
about the results (repeat and autocomplete were not turned on).

At some point I loaded the game, closed the tab without saving, and when I
opened it again, the load game option was greyed out.

**Comment by frnsys** [comment]
Hm alright. Are you able to inspect your browser's local storage when you load the game site? I'm wondering if it's a problem of the game not having saved at all to local storage or if it _was_ saved but in a malformed way.

**Comment by a-horde-of-cats** [comment]
I am, though I might have moved on a couple save files since then.. hopefully there's some remnant or error log somewhere I can get for you. Filepath?

**Comment by frnsys** [comment]
sadly those errors logs aren't saved to disk...but if it does happen again, open the javascript console and copy any messages that appear there and I'll take a closer look.

**Comment by a-horde-of-cats** [comment]
This isn't about the savefile loss problem, but the "projects stop notifying on completion" bug; see attached log. While this log was collected I completed a couple of products with no window popping up for product type or build.
[thefounder.biz-1598927693045.log](https://github.com/frnsys/the_founder/files/5153616/thefounder.biz-1598927693045.log)

**Comment by a-horde-of-cats** [comment]
This log was when I tried to fix the issue by deleting the jobs, but some of the events didn't disappear
[thefounder.biz-1598928673943.log](https://github.com/frnsys/the_founder/files/5153663/thefounder.biz-1598928673943.log)

**Comment by frnsys** [comment]
Great!

For `thefounder.biz-1598928673943.log`, best as I can tell from the log is that there is a bug with stopping a task, and that broke things down the line. I'll add a quick fix that should prevent that from breaking other things, though I'm not sure what the deeper cause of the error is.

For `thefounder.biz-1598927693045.log` it seems to be some issue with employees leaving objects they're interacting with...again, don't know the deeper cause (been awhile since I looked at this code!) but I think this patch should at least stop it from breaking other parts of the game.

I'll deploy the changes shortly. Thanks for collecting these logs.

**Comment by frnsys** [comment]
Alright, the game should be updated. I'll close this issue for now but feel free to re-open it if you're still having this problem.

**Comment by a-horde-of-cats** [comment]
I'm sorry to say the problems with projects not generating a popup continue. Deleting all ongoing projects, saving, then restarting fixes it for a few minutes. Also, research projects complete without a notification more often than not, even if products are working.

[thefounder.biz-1598988266850.log](https://github.com/frnsys/the_founder/files/5158641/thefounder.biz-1598988266850.log)
[thefounder.biz-1598992066031.log](https://github.com/frnsys/the_founder/files/5158642/thefounder.biz-1598992066031.log)

I got two logs here. One was of a game where tasks were disappearing without notification as before (the idle workers count would just go up), and another log is a similar situation but the game had a crash as I was recording.

Sorry for the bother and thanks for coming back and trying to read code from the distant past for me. I'm guessing some Chrome update or similar broke something important in the time since the game was released..

**Comment by frnsys** [comment]
Huh strange--those logs look good, I don't see any errors related to the game there. There is a request for `undefined.png` which looks suspicious though...this may take more time than I have right now, but if I'll try to look into it more deeply.

**Comment by a-horde-of-cats** [comment]
Fair enough! Again, thank you for the time you've already put in supporting a past project.

**Comment by a-horde-of-cats** [comment]
You will be pleased to know that I potentially have a fix for the bug.

I was (on the advice of the person who introduced me to the game) having 2 or more products on repeat at all times for income. Each product combo I discovered was made hundreds of times and the revenue feed was endlessly clogged with e.g. yPhone 253.

It felt pretty unrealistic, but I'm pretty sure it was making the game struggle in some way or another.

Made a new game file called NoRepeats Inc. I stopped putting products on repeat and reserved the function for promos. No bug.

It's not my place to make game design suggestions, but I believe the game released without the repeat function. Maybe remove it from products and leave it available on promos; probably fits your original vision better anyway.

**Comment by frnsys** [comment]
Thanks for looking more into it--I'm trying to see what might be causing the issue...do you also have tasks set to autocomplete? Wondering if there is some conflict going on there with the task repeating.

**Comment by frnsys** [comment]
Maybe this will help narrow down the problem. When a product task is finished and repeated, the following should happen (in `Task.finish`):

- The product launches (`Product.launch`), which basically triggers `Product.onProductLaunch` which is set in `Manage.create`.
    - The game is paused
    - The product designer view is shown (in `Manage.enterTheMarket`)
    - Once the product designer view is done, the market mini game starts
- The old task is copied
- The company starts the new task copy

At a high level I think that's everything that triggers after a product task is finished.

Just to see if I understand correctly, the bug you're experiencing is where, after finishing the product task:

- the product designer view no longer shows up
- the product task no longer repeats

**Comment by a-horde-of-cats** [comment]
I haven't tested in all situations, but when the bugs show up, there is always lots of projects in the revenue tracker on the left, and projects have been set on repeat and to finish automatically.

The bug involves non-repeating, non-auto-finish projects when the gamestate is as described above.

>after finishing the product task:
>the product designer view no longer shows up

Yes. The progress circle disappears but the product designer view doesn't pop up. Completion notifications also don't appear for research and events, or sometimes appear very delayed.

>the product task no longer repeats

_SOMETIMES?_
Some repeating products will stop repeating and need to be deleted; the new replacement task is made but the progress bar doesn't advance.
However, new products that I try to make without repeating will just not complete and be gone from the list of in-progress projects.

**Comment by frnsys** [comment]
Interesting, this seems a lot more complicated than I originally anticipated...when I have a bit more time I'll see if I can recreate it on my end and see where that gets me.

Thanks again for your help on this.

**Comment by a-horde-of-cats** [comment]
Thanks for the reply and for reopening, hope you manage to fix it. I have finally finished a mostly bug-free playthrough of the game (by not using the repeat function on products) so be assured it is not a bug that makes the game unplayable.
